src/main/java/net/mcreator/ui/minecraft/states/item/JItemPropertiesStatesList.java

```diff
@@ -0,0 +1,269 @@
+import net.mcreator.ui.validation.validators.UniqueNameValidator;
+
+import javax.annotation.Nullable;
+import javax.swing.*;
+import java.awt.*;
+import java.awt.event.ContainerAdapter;
+import java.awt.event.ContainerEvent;
+import java.util.List;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class JItemPropertiesStatesList extends JEntriesList {
+
+	private final List<JItemPropertiesListEntry> propertiesList = new ArrayList<>();
+	private final List<JItemStatesListEntry> statesList = new ArrayList<>();
+	private final AtomicInteger propertyId = new AtomicInteger(0);
+
+	private final List<String> builtinPropertyNames;
+	private final Map<String, BuiltInPropertyData<?>> builtinProperties = new LinkedHashMap<>();
+
+	private final JPanel propertyEntries = new JPanel() {
+		@Override public void setEnabled(boolean enabled) {
+			super.setEnabled(enabled);
+			propertiesList.forEach(e -> e.setEnabled(enabled));
+		}
+	};
+	private final JPanel stateEntries = new JPanel() {
+		@Override public void setEnabled(boolean enabled) {
+			super.setEnabled(enabled);
+			statesList.forEach(e -> e.setEnabled(enabled));
+		}
+	};
+
+	private final TechnicalButton addProperty = new TechnicalButton(UIRES.get("16px.add.gif"));
+	private final TechnicalButton addState = new TechnicalButton(UIRES.get("16px.add.gif"));
+
+	public JItemPropertiesStatesList(MCreator mcreator, IHelpContext gui) {
+		super(mcreator, new GridLayout(), gui);
+		setOpaque(false);
+		setBorder(BorderFactory.createEmptyBorder(5, 10, 0, 10));
+
+		Map<String, DataListEntry> properties = DataListLoader.loadDataMap("itemproperties");
+		builtinPropertyNames = List.copyOf(properties.keySet());
+
+		PropertyData.Logic logic = new PropertyData.Logic(""); // needed for hardcoded logic properties
+		for (DataListEntry entry : properties.values()) {
+			PropertyData.FloatNumber builtin;
+			if ("Number".equals(entry.getType()) && entry.getOther() instanceof Map<?, ?> other) {
+				float min = Float.parseFloat((String) other.get("min"));
+				float max = Float.parseFloat((String) other.get("max"));
+				builtin = new PropertyData.FloatNumber(entry.getName(), min, max);
+			} else if ("Logic".equals(entry.getType())) {
+				builtin = new PropertyData.FloatNumber(entry.getName(), 0, 1) {
+					@Override public JComponent getComponent(MCreator mcreator, @Nullable Object value) {
+						return logic.getComponent(mcreator, value != null && (Float) value == 1);
+					}
+
+					@Override public Float getValue(JComponent component) {
+						return logic.getValue(component) ? 1 : 0f;
+					}
+				};
+			} else {
+				continue;
+			}
+			builtinProperties.put(entry.getName(), new BuiltInPropertyData<>(builtin));
+		}
+
+		propertyEntries.setLayout(new GridLayout(0, 1, 5, 5));
+		propertyEntries.setOpaque(false);
+		stateEntries.setLayout(new GridLayout(0, 1, 5, 5));
+		stateEntries.setOpaque(false);
+
+		propertyEntries.addContainerListener(new ContainerAdapter() {

```

Comment:
Given you control when entry is removed, there is no need to hijack ContainerListener, but simply do things when remove button is pressed

Reply:
There is no <kbd>Remove</kbd> button listener passed to child entry constructor, so I can either add a container listener to the entry panel or override a method in the entry list.

Reply:
States list entry can hold reference to parent (JItemPropertiesStatesList) and call method to said parent

Reply:
Oh, wait, given it no longer uses `PanelUtils.expandHorizontally()`... The listener should not be that hard to do.

Reply:
You should not use listener but just call a method.

> JItemPropertiesListEntry can hold reference to parent (JItemPropertiesStatesList) and call method to said parent when JItemPropertiesListEntry is being removed

Reply:
...
Not saying anything against this, but this is what I was trying to avoid (looking at other list entries not having access to the holding list panel.

Reply:
✔️ Fixed